source/module_basis/module_nao/hydrogen_radials.cpp

```diff
@@ -0,0 +1,412 @@
+                                    const std::vector<double> rvalue)
+{
+    std::vector<double> integrand(rvalue.size());
+    for(int ir = 0; ir != rvalue.size(); ++ir)
+    {
+        integrand[ir] = rvalue[ir] * rvalue[ir] * rgrid[ir] * rgrid[ir];
+    }
+    double dr = rgrid[1] - rgrid[0];
+    double norm = ModuleBase::Integral::simpson(rvalue.size(), integrand.data(), dr);
+    norm = sqrt(norm);
+    return norm;
+}
+
+double HydrogenRadials::generate_hydrogen_radial_toconv(const double charge,
+                                                          const int n,
+                                                          const int l,
+                                                          const double conv_thr,
+                                                          const int rank,
+                                                          std::vector<double>& rgrid,
+                                                          std::vector<double>& rvalue,
+                                                          std::ofstream* ptr_log)
+{
+    double norm = 0.0;
+    double rmax_ = 0.0; // in Bohr
+    double rmin_ = 0.0; // always to be 0, in Bohr
+    double delta_r = 0.5; // stepsize for radius cutoff searching, in Bohr
+
+    // clear the input vectors
+    rgrid.clear(); rgrid.shrink_to_fit();
+    rvalue.clear(); rvalue.shrink_to_fit();
+    double dr = 0.01; // radial function realspace grid stepsize, in Bohr
+    if(delta_r < dr)
+    {
+        dr = delta_r;
+    }
+    printf("Searching for the cutoff radius for n = %d, l = %d, conv_thr = %6.4e\n", n, l, conv_thr);
+    printf("%10s%12s%14s%18s", "Step Nr.", "Rmax (a.u.)", "Norm", "Delta Norm\n");
+    int istep = 1;
+    double delta_norm = 1.0;
+    while((std::fabs(delta_norm) > conv_thr))
+    {
+        rmin_ = rmax_;
+        rmax_ += delta_r;
+        int ngrid = static_cast<int>((rmax_ - rmin_) / dr) + 1; // [rmin, rmax]
+        std::vector<double> rgrid_segment(ngrid);
+        for(int ir = 0; ir != ngrid; ++ir)
+        {
+            rgrid_segment[ir] = rmin_ + ir * dr;
+        }
+        std::vector<double> rvalue_segment = generate_hydrogen_radial_segment(
+            charge, n, l, rmin_, rmax_, dr, rank, ptr_log);
+        // before push back, pop back the last element
+        if(rgrid.size() != 0)
+        {
+            rgrid.pop_back();
+            rvalue.pop_back();
+        }
+        rgrid.insert(rgrid.end(), rgrid_segment.begin(), rgrid_segment.end());
+        rvalue.insert(rvalue.end(), rvalue_segment.begin(), rvalue_segment.end());
+        delta_norm = norm;
+        norm = radial_norm(rgrid, rvalue);
+        delta_norm = norm - delta_norm;
+        if(istep == 0) printf("%10d%12.2f%14.10f%18.10e\n", istep, rmax_, norm, delta_norm);
+        ++istep;
+    }
+    printf("%10d%12.2f%14.10f%18.10e\n", istep, rmax_, norm, delta_norm);
+    return rmax_;
+}
+
+std::vector<std::pair<int, int>> HydrogenRadials::unzip_strategy(const int nmax,
```

Comment:
strategy keyword intrinsically supports three options, minimal, energy and full. For each strategy, generate different (n, l) pair, n is principle quantum number and l is angular momentum number. According to these pairs, call generate hydrogen-like orbitals, with the function above.